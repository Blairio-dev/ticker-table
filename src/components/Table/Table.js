import React, { useState } from "react";
import PropTypes from "prop-types";
import "./Table.scss";
import { isNumberPositive } from "../../utilities/functions";

const getPriceClassNameSuffix = (price) =>
  isNumberPositive(price) ? "positive" : "negative";

const headerNames = {
  assetClass: "Asset Class",
  price: "Price",
  ticker: "Ticker",
};

const sortAlphabetically = (data) => [
  ...data.sort(function (a, b) {
    const x = a.ticker.toLowerCase();
    const y = b.ticker.toLowerCase();
    if (x < y) {
      return -1;
    }
    if (x > y) {
      return 1;
    }
    return 0;
  }),
];

const sortNumerically = (data) => [
  ...data.sort(function (a, b) {
    return b.price - a.price;
  }),
];

const sortByAssetClass = (data) => {
  const creditAssets = data.filter((asset) => asset.assetClass === "Credit");
  const equityAssets = data.filter((asset) => asset.assetClass === "Equities");
  const macroAssets = data.filter((asset) => asset.assetClass === "Macro");

  return equityAssets.concat(macroAssets).concat(creditAssets);
};

const onClickHandler = (key, data) => {
  switch (key) {
    case "ticker":
      return sortAlphabetically(data);
    case "price":
      return sortNumerically(data);
    case "assetClass":
      return sortByAssetClass(data);
    default:
      return data;
  }
};

const Table = ({ data }) => {
  const [sortedData, sortData] = useState(data);
  const keys = Object.keys(data[0]);
  return (
    <table>
      <caption>Financial Instruments</caption>
      <thead>
        <tr>
          {keys.map(
            (key) =>
              headerNames[key] && (
                <th key={key}>
                  <button
                    onClick={() => sortData(onClickHandler(key, sortedData))}
                  >
                    {headerNames[key]}
                  </button>
                </th>
              )
          )}
        </tr>
      </thead>
      <tbody>
        {sortedData.map((asset) => (
          <tr className={`assetClass-${asset.assetClass}`} key={asset.ticker}>
            <td>{asset.ticker}</td>
            <td className={`price-${getPriceClassNameSuffix(asset.price)}`}>
              {asset.price}
            </td>
            <td>{asset.assetClass}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

Table.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      ticker: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      assetClass: PropTypes.string.isRequired,
    })
  ),
};

export { Table };
