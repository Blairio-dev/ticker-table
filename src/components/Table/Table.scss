@import "./colours.scss";

table {
  border-spacing: 0;
  text-align: left;
  width: 100%;
}

caption {
  padding: 0 16px 8px 0;
  text-align: inherit;
}

thead {
  background: $black;
}

tbody {
  overflow: scroll;

  tr {
    &:hover {
      box-shadow: 0 0 4px $boxShadow;
      transform: scale(1.005);

      td,
      th {
        border-color: $borderActive;
      }
    }
  }
}

tr {
  transition: 0.1s ease-out;
}

th,
td {
  border-collapse: collapse;
  white-space: nowrap;
}

th {
  &:hover {
    button {
      border-color: $borderActive;
    }
  }

  button {
    background: transparent;
    border: 1px solid $border;
    color: $textLight;
    cursor: pointer;
    font-size: inherit;
    text-align: inherit;
    width: 100%;

    &:focus {
      border-color: $borderActive;
      outline: none;
    }
  }
}

td {
  border: 1px solid $border;
}

td,
button {
  padding: 8px 16px;

  @media (max-width: 960px) {
    padding: 8px 8px;
  }
}

.assetClass-Credit {
  background-color: $green;
}

.assetClass-Equities {
  background-color: $blue;
}

.assetClass-Macro {
  background-color: $white;
  color: $textDark;
}

.price-negative {
  background-color: $negative;
  color: $textLight;
}

.price-positive {
  background-color: $positive;
  color: $textLight;
}
