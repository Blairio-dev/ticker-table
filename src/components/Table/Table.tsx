import { useState } from "react";
import "./Table.scss";
import {
  isNumberPositive,
  sortAlphabetically,
  sortByAssetClass,
  sortNumerically,
} from "../../assets/functions";
import { headerNames, Instrument } from "../../assets/interfaces";
import { ReactComponent as SortIcon } from "../../assets/icons/sort.svg";

const getPriceClassNameSuffix = (price: number) =>
  isNumberPositive(price) ? "positive" : "negative";

const onClickHandler = (key: string, data: Instrument[]) => {
  switch (key) {
    case "ticker":
      return sortAlphabetically(data);
    case "price":
      return sortNumerically(data);
    case "assetClass":
      return sortByAssetClass(data);
    default:
      return data;
  }
};

interface TableProps {
  data: Instrument[];
}

const Table = ({ data }: TableProps) => {
  const [sortedData, sortData] = useState(data);
  const keys = Object.keys(data[0]);
  return (
    <table>
      <caption>Financial Instruments</caption>
      <thead>
        <tr>
          {keys.map(
            (key) =>
              headerNames[key] && (
                <th key={key}>
                  <button
                    onClick={() => sortData(onClickHandler(key, sortedData))}
                  >
                    <div className="inline-wrapper">
                      {headerNames[key]}
                      <SortIcon />
                    </div>
                  </button>
                </th>
              )
          )}
        </tr>
      </thead>
      <tbody>
        {sortedData.map((asset) => (
          <tr className={`assetClass-${asset.assetClass}`} key={asset.ticker}>
            <td>{asset.ticker}</td>
            <td className={`price-${getPriceClassNameSuffix(asset.price)}`}>
              {asset.price}
            </td>
            <td>{asset.assetClass}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export { Table };
