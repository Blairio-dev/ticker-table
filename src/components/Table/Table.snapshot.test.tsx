import React from "react";
import renderer from "react-test-renderer";
import { Table } from "./Table";

const data = [
  {
    ticker: "UNO",
    price: 1111.11,
    assetClass: "Credit",
    unsupportedKey: "test",
  },
  {
    ticker: "DOS",
    price: 2222.22,
    assetClass: "Equities",
  },
  {
    ticker: "TRES",
    price: 3333.33,
    assetClass: "Macro",
  },
  {
    ticker: "UN",
    price: -1111.11,
    assetClass: "Credit",
  },
  {
    ticker: "DEUX",
    price: -2222.22,
    assetClass: "Equities",
  },
  {
    ticker: "TROIS",
    price: -333.33,
    assetClass: "Macro",
  },
];

const DEFAULT_PROPS = {
  data,
};

describe("DataTable Snapshots", () => {
  it("Renders with default props", () => {
    const tree = renderer.create(<Table {...DEFAULT_PROPS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
