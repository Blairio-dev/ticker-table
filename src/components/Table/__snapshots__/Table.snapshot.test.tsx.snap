// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Snapshots Renders with default props 1`] = `
<table>
  <caption>
    Financial Instruments
  </caption>
  <thead>
    <tr>
      <th>
        <button
          onClick={[Function]}
        >
          <div
            className="Inline-wrapper"
          >
            <div>
              Ticker
            </div>
            <div>
              <svg>
                sort.svg
              </svg>
            </div>
          </div>
        </button>
      </th>
      <th>
        <button
          onClick={[Function]}
        >
          <div
            className="Inline-wrapper"
          >
            <div>
              Price
            </div>
            <div>
              <svg>
                sort.svg
              </svg>
            </div>
          </div>
        </button>
      </th>
      <th>
        <button
          onClick={[Function]}
        >
          <div
            className="Inline-wrapper"
          >
            <div>
              Asset Class
            </div>
            <div>
              <svg>
                sort.svg
              </svg>
            </div>
          </div>
        </button>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="assetClass-Credit"
    >
      <td>
        UNO
      </td>
      <td
        className="price-positive"
      >
        1111.11
      </td>
      <td>
        Credit
      </td>
    </tr>
    <tr
      className="assetClass-Equities"
    >
      <td>
        DOS
      </td>
      <td
        className="price-positive"
      >
        2222.22
      </td>
      <td>
        Equities
      </td>
    </tr>
    <tr
      className="assetClass-Macro"
    >
      <td>
        TRES
      </td>
      <td
        className="price-positive"
      >
        3333.33
      </td>
      <td>
        Macro
      </td>
    </tr>
    <tr
      className="assetClass-Credit"
    >
      <td>
        UN
      </td>
      <td
        className="price-negative"
      >
        -1111.11
      </td>
      <td>
        Credit
      </td>
    </tr>
    <tr
      className="assetClass-Equities"
    >
      <td>
        DEUX
      </td>
      <td
        className="price-negative"
      >
        -2222.22
      </td>
      <td>
        Equities
      </td>
    </tr>
    <tr
      className="assetClass-Macro"
    >
      <td>
        TROIS
      </td>
      <td
        className="price-negative"
      >
        -333.33
      </td>
      <td>
        Macro
      </td>
    </tr>
  </tbody>
</table>
`;
